From: Felipe Borges <felipeborges@gnome.org>
Date: Tue, 19 Mar 2024 15:34:32 +0100
Subject: build, privacy: Make Location settings build conditional

Considering the retirement of Mozilla Location Service, most
laptop/desktop users will rely on weaker network based location
sources.

Let's make the Location settings a build option, disabled by default.
-Dlocation-services=enabled

See #2959

(cherry picked from commit a773c61b64d07adfdfb7fce49a6fbe877e6141f1)

Origin: future 46.1
---
 meson.build                        |  5 +++++
 meson_options.txt                  |  1 +
 panels/privacy/cc-privacy-panel.c  | 11 +++++++++++
 panels/privacy/cc-privacy-panel.ui |  7 ++-----
 4 files changed, 19 insertions(+), 5 deletions(-)

diff --git a/meson.build b/meson.build
index 547b42b..2e10348 100644
--- a/meson.build
+++ b/meson.build
@@ -274,6 +274,11 @@ endif
 config_h.set('HAVE_MALCONTENT', enable_malcontent,
              description: 'Define to 1 if malcontent support is enabled')
 
+# location services support
+location_services = get_option('location-services')
+config_h.set('HAVE_LOCATION_SERVICES', location_services.enabled(),
+             description: 'Whether location services is enabled')
+
 if host_is_linux
   # ModemManager
   mm_dep =  dependency('mm-glib', version: '>= 0.7')
diff --git a/meson_options.txt b/meson_options.txt
index e53d6e3..f415a7e 100644
--- a/meson_options.txt
+++ b/meson_options.txt
@@ -1,5 +1,6 @@
 option('deprecated-declarations', type: 'feature', value: 'disabled', description: 'build with deprecated declaration warnings')
 option('documentation', type: 'boolean', value: false, description: 'build documentation')
+option('location-services', type: 'feature', value: 'disabled', description: 'build with location services')
 option('ibus', type: 'boolean', value: true, description: 'build with IBus support')
 option('privileged_group', type: 'string', value: 'wheel', description: 'name of group that has elevated permissions')
 option('snap', type: 'boolean', value: true, description: 'build with Snap support')
diff --git a/panels/privacy/cc-privacy-panel.c b/panels/privacy/cc-privacy-panel.c
index 1d16a78..d7cfd9c 100644
--- a/panels/privacy/cc-privacy-panel.c
+++ b/panels/privacy/cc-privacy-panel.c
@@ -41,6 +41,7 @@ struct _CcPrivacyPanel
 
   AdwNavigationView *navigation;
   CcListRow         *bolt_row;
+  CcListRow         *location_row;
 };
 
 CC_PANEL_REGISTER (CcPrivacyPanel, cc_privacy_panel)
@@ -73,6 +74,7 @@ cc_privacy_panel_class_init (CcPrivacyPanelClass *klass)
 
   gtk_widget_class_bind_template_child (widget_class, CcPrivacyPanel, navigation);
   gtk_widget_class_bind_template_child (widget_class, CcPrivacyPanel, bolt_row);
+  gtk_widget_class_bind_template_child (widget_class, CcPrivacyPanel, location_row);
 
   g_type_ensure (CC_TYPE_CAMERA_PAGE);
   g_type_ensure (CC_TYPE_DIAGNOSTICS_PAGE);
@@ -114,5 +116,14 @@ cc_privacy_panel_init (CcPrivacyPanel *self)
                           self->bolt_row, "visible", G_BINDING_SYNC_CREATE);
 #endif
 
+#ifdef HAVE_LOCATION_SERVICES
+  CcLocationPage *location_page = g_object_new (CC_TYPE_LOCATION_PAGE, NULL);
+
+  adw_navigation_view_add (self->navigation, ADW_NAVIGATION_PAGE (location_page));
+
+  g_object_bind_property (location_page, "visible",
+                          self->location_row, "visible", G_BINDING_SYNC_CREATE);
+#endif
+
   g_signal_connect_object (self, "notify::subpage", G_CALLBACK (on_subpage_set), self, G_CONNECT_SWAPPED);
 }
diff --git a/panels/privacy/cc-privacy-panel.ui b/panels/privacy/cc-privacy-panel.ui
index 8ba3c2c..5b216bf 100644
--- a/panels/privacy/cc-privacy-panel.ui
+++ b/panels/privacy/cc-privacy-panel.ui
@@ -32,7 +32,8 @@
                         </child>
 
                         <child>
-                          <object class="CcListRow">
+                          <object class="CcListRow" id="location_row">
+                            <property name="visible">False</property>
                             <property name="title" translatable="yes">_Location</property>
                             <property name="subtitle" translatable="yes">Control access to your location</property>
                             <property name="icon-name">location-access-symbolic</property>
@@ -136,10 +137,6 @@
           <object class="CcScreenPage"/>
         </child>
 
-        <child>
-          <object class="CcLocationPage"/>
-        </child>
-
         <child>
           <object class="CcCameraPage"/>
         </child>
